### PR TITLE
feat(app): Made API URL configurable in app

### LINKED ISSUE #584: Make API base URL configurable in browser application
### Overview

Currently, the base URL for the API is hardcoded to `http://localhost:4040/api` in the front-end application. However, the API's URL should be configurable.

Update to add context https://github.com/hackforla/HomeUniteUs/issues/584#issuecomment-1707579200 (thanks @Joshua-Douglas):

> This issue was necessary to properly use our build-deploy-ec2.yml and build-dev-workflow.yaml workflows. If the application is deployed with the hard coded URL then some features of the deployed app, such as sign in & sign up, will not work.

### Action Items

- [x] Update the API's base URL in `api.ts` to be obtained by an environment variable (e.g. from a `.env` file). The variable can be written as `import.meta.env.VITE_HUU_API_BASE_URL`.
- [x] Add an entry to `vite-en.d.ts` with `readonly VITE_HUU_API_BASE_URL: string;`.
- [x] Add `VITE_HUU_API_BASE_URL=http://localhost:4040/api` to `.env.example`.
- [x] Add `VITE_HUU_API_BASE_URL` as a GitHub environment variable (Tyler or Erik) 

### Resources/Instructions

[Vite environment variables for production](https://vitejs.dev/guide/env-and-mode.html#production-replacement)

Related #502

### PR DESCRIPTION
**Pull request by paulespinosa**
Closes #584

feat(app): Made API URL configurable in app

This commit introduces a new environment variable: VITE_HUU_API_BASE_URL. This
environment variable represents the base URL of the API that the app should
contact.

During development, if this environment variable is not set, then the API's base
URL will be set to "http://localhost:4040/api/" automatically. This is also true
when running tests via `vitest` aka `npn run test`. `vitest` enables the `test` mode.

During a non-development build, if the environment variable is not set, then the
build will throw an error. It is required for non-development builds.

The app's README has been updated to explain configuration and production builds.

The GitHub build workflow has been updated to pull in the environment variables from
GitHub environment value settings.

The GitHub workflows have been refactored. One workflow just runs tests. The
other runs tests, builds, and deploys.

The ForgotPasswordCode test has been updated so that the mock request
interceptor performs a regex matches against the request URL.

Non-deterministic test failures were encountered during development. The
symptoms were unhandled rejection errors. The test code was missing await
statements that would sometimes cause futures to complete after the tests
finished running.

Another set of test errors were encountered. The test code was checking for a
"form" role on the page. However, HTML forms are not assigned a role that
`testing-library` can see.

Due to the test errors above, ForgotPasswordCode.test.tsx and
ResetPassword.test.tsx have been updated to fix the test errors.

### CODE REVIEW

**Comment by erikguntner** [comment]
@paulespinosa this all looks good to me! I tried adding VITE_API_BASE_URL as a secret and as an environment variable and rerunning the checks, but they still failed with the same error `Error: VITE_API_BASE_URL is not configured`. I'm wondering if we have to do something like they've done [here](https://github.com/facebook/create-react-app/discussions/9064#discussioncomment-30344) and set this as a secret and reference it in our workflow.

**Review by Joshua-Douglas — CHANGES_REQUESTED** [review]
Hey @paulespinosa and @erikguntner, 

Are we confident that we need to store the API URL as a github secret?

I did some digging into the [multi-environment configuration options provided by react](https://create-react-app.dev/docs/adding-custom-environment-variables/#what-other-env-files-can-be-used) and I think we can inject the API URL as a environment variable using the `.env` files. 

I demonstrated this by created three separate .env files:


```.env
# .env.development
VITE_API_URL=http://localhost:8080
```

```.env
# .env.staging
VITE_API_URL=http://dev.homeunite.us
```

```.env
#.env.production
VITE_API_URL=http://homeuniteus.us
```

And then updated the npm package.json settings to look like this:
```
    "dev": "vite --mode development",
    "build:staging": "tsc && vite build --mode staging",
    "build:production": "tsc && vite build --mode production",
    "preview": "vite preview",
```

With this configuration I was able to access the API URL within our javascript using `import.meta.env.VITE_API_URL`. I build the app using the three new configurations and used the console log to verify the API URL.

```
# Test 1:
npm run dev  # VITE_API_URL returns http://localhost:8080
# Test 2:
npm run build:staging
npm run preview # VITE_API_URL returns http://dev.homeunite.us
# Test 3:
npm run build:production
npm run preview http://homeunite.us
````

What do you two think? Would this work?

File: app/vite.config.ts [hunk 1]
```diff
@@ -1,25 +1,44 @@
...
+function apiBaseUrl(envApiBaseUrl: string, mode: string): string | never {
+  const url = JSON.stringify(envApiBaseUrl);
+  if (url) {
+    return url;
+  } else {
+    if (mode == 'development') {
+      return JSON.stringify('http://localhost:4040/api/');
+    } else {
+      throw new Error('VITE_API_BASE_URL is not configured');
+    }
```
[the remark this answers is not in the document]
Comment (paulespinosa) [reply app/vite.config.ts:16]: It does execute. But, your refactoring suggestings will make it better. 

The build runner shows the error message:

https://github.com/hackforla/HomeUniteUs/actions/runs/6040992137/job/16392991890#step:5:11

>  tsc && vite build
> 
> failed to load config from /home/runner/work/HomeUniteUs/HomeUniteUs/app/vite.config.ts
> error during build:
> Error: VITE_API_BASE_URL is not configured
>     at apiBaseUrl (/home/runner/work/HomeUniteUs/HomeUniteUs/app/vite.config.ts:42:13)
>     at /home/runner/work/HomeUniteUs/HomeUniteUs/app/vite.config.ts:50:25
>     at loadConfigFromFile (file:///home/runner/work/HomeUniteUs/HomeUniteUs/app/node_modules/vite/dist/node/chunks/dep-2faf2534.js:63023:15)
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
>     at async resolveConfig (file:///home/runner/work/HomeUniteUs/HomeUniteUs/app/node_modules/vite/dist/node/chunks/dep-2faf2534.js:62643:28)
>     at async doBuild (file:///home/runner/work/HomeUniteUs/HomeUniteUs/app/node_modules/vite/dist/node/chunks/dep-2faf2534.js:45631:20)
>     at async build (file:///home/runner/work/HomeUniteUs/HomeUniteUs/app/node_modules/vite/dist/node/chunks/dep-2faf2534.js:45620:16)
>     at async CAC.<anonymous> (file:///home/runner/work/HomeUniteUs/HomeUniteUs/app/node_modules/vite/dist/node/cli.js:756:9)
> Error: Process completed with exit code 1.

**Comment by paulespinosa** [comment]
Hi @Joshua-Douglas, thanks for helping out on this one.

> Are we confident that we need to store the API URL as a github secret?

It shouldn't be a secret. It should be a GitHub Environment configuration instead. https://docs.github.com/en/actions/learn-github-actions/variables#creating-configuration-variables-for-an-environment


> 
> I did some digging into the [multi-environment configuration options provided by react](https://create-react-app.dev/docs/adding-custom-environment-variables/#what-other-env-files-can-be-used) and I think we can inject the API URL as a environment variable using the `.env` files.
> 
> I demonstrated this by created three separate .env files:
>
> What do you two think? Would this work? If we opt for this approach then we could remove the custom configuration code introduced w/in this PR and just rely on `import.meta.env.VITE_API_URL` instead.

Keep in mind that the GitHub build runner doesn't have the .env files.

**Comment by paulespinosa** [comment]
@erikguntner @Joshua-Douglas we can discuss dotenv options on the issue @ju1es created: https://github.com/hackforla/HomeUniteUs/issues/529#issuecomment-1702097176

**Comment by paulespinosa** [comment]
> Hey @paulespinosa and @erikguntner,
> 
> Are we confident that we need to store the API URL as a github secret?
> 
> I did some digging into the [multi-environment configuration options provided by react](https://create-react-app.dev/docs/adding-custom-environment-variables/#what-other-env-files-can-be-used) and I think we can inject the API URL as a environment variable using the `.env` files.
> 
> I demonstrated this by created three separate .env files:
> 
> ```dotenv
> # .env.development
> VITE_API_URL=http://localhost:8080
> ```
> 
> ```dotenv
> # .env.staging
> VITE_API_URL=http://dev.homeunite.us
> ```
> 
> ```dotenv
> #.env.production
> VITE_API_URL=http://homeuniteus.us
> ```
> 
> And then updated the npm package.json settings to look like this:
> 
> ```
>     "dev": "vite --mode development",
>     "build:staging": "tsc && vite build --mode staging",
>     "build:production": "tsc && vite build --mode production",
>     "preview": "vite preview",
> ```
> 
> With this configuration I was able to access the API URL within our javascript using `import.meta.env.VITE_API_URL`. I build the app using the three new configurations and used the console log to verify the API URL.
> 
> ```
> # Test 1:
> npm run dev  # VITE_API_URL returns http://localhost:8080
> # Test 2:
> npm run build:staging
> npm run preview # VITE_API_URL returns http://dev.homeunite.us
> # Test 3:
> npm run build:production
> npm run preview http://homeunite.us
> ```
> 
> What do you two think? Would this work? If we opt for this approach then we could remove the custom configuration code introduced w/in this PR and just rely on `import.meta.env.VITE_API_URL` instead.

Since the variable is required, it seems that the build should fail if it doesn't exist. If the `VITE_API_BASE_URL` environment is not defined then the build will replace `import.meta.env.VITE_API_BASE_URL`  the value with `{}.VITE_API_BASE_URL` without error. Then issues would be experienced by clients.

**Comment by paulespinosa** [comment]
> @paulespinosa this all looks good to me! I tried adding VITE_API_BASE_URL as a secret and as an environment variable and rerunning the checks, but they still failed with the same error `Error: VITE_API_BASE_URL is not configured`. I'm wondering if we have to do something like they've done [here](https://github.com/facebook/create-react-app/discussions/9064#discussioncomment-30344) and set this as a secret and reference it in our workflow.

@erikguntner thank you! I'll update the workflows to explicitly reference the variable.

**Comment by Joshua-Douglas** [comment]
> Keep in mind that the GitHub build runner doesn't have the .env files.

I see your point. Yeah, I was thinking we could commit `.env.staging`, `.env.development` and `.env.production` to contain the different environment variables. That should be safe as long they only contain variables, but after reading your link I agree that specifying these variables within separate GitHub environments could be cleaner.

I think what confused me is that your code is explicitly loading a `.env` file from the current working directory. The vite `loadEnv` file will use the mode and directory to search for the correct env file to load the variables from (i.e. `vite build --mode staging` will try to load `.env.staging`). How will the build runner get an .env file if we aren't committing it? Do we plan to generate one using a script?

**Comment by paulespinosa** [comment]
> I think what confused me is that your code is explicitly loading a `.env` file from the current working directory. The vite `loadEnv` file will use the mode and directory to search for the correct env file to load the variables from (i.e. `vite build --mode staging` will try to load `.env.staging`). How will the build runner get an .env file if we aren't committing it? Do we plan to generate one using a script?

Yes, it is confusing. [`loadEnv` uses `dotenv`](https://v3.vitejs.dev/guide/api-javascript.html#loadenv). One of the subtlety mentioned features of `dotenv` is that it will first load in the system environment variables.

In the example below, the environment variable is set/exported. It shows that the `.env` files do not exist in the directory. `npm run build` is run and the results are inspected for the value of the environment variable.

```shell
[~/Development/HackForLA/HomeUniteUs/app]$ export VITE_API_BASE_URL=https://example.com/api
[~/Development/HackForLA/HomeUniteUs/app]$ ls -a .env*           584-app-configurable-api-url
zsh: no matches found: .env*
[~/Development/HackForLA/HomeUniteUs/app]$ npm run build         584-app-configurable-api-url

> home-unite-us@1.0.0 build
> tsc && vite build

vite v3.2.7 building for production...
✓ 11774 modules transformed.
dist/assets/favicon.17e50649.svg   1.49 KiB
dist/assets/spy.4822d67d.png       4.79 KiB
dist/assets/favicon.12d83c1c.png   11.79 KiB
dist/index.html                    0.77 KiB
dist/assets/index.5265c558.css     0.29 KiB / gzip: 0.22 KiB
dist/assets/index.4535eec7.js      588.31 KiB / gzip: 188.63 KiB

(!) Some chunks are larger than 500 KiB after minification. Consider:
- Using dynamic import() to code-split the application
- Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/guide/en/#outputmanualchunks
- Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.

[~/Development/HackForLA/HomeUniteUs/app]$ js-beautify dist/assets/index.4535eec7.js | grep "example\.com"
            baseUrl: "https://example.com/api",
```

**Comment by paulespinosa** [comment]
@erikguntner @Joshua-Douglas I made updates the PR. The `Run cd app && npm run build` step in the GitHub workflow is successful! However, it looks like the URL is missing the path `/api/`.

But the `Run cd app && npm run test` step fails on `ForgotPasswordCode.test.tsx`. It looks like it is an integration test. I'll try to copy the mocking technique from the `ResetPassword.test.tsx` file.

**Comment by paulespinosa** [comment]
@erikguntner @Joshua-Douglas The code has been updated again. The variable name has been changed to `VITE_HUU_API_BASE_URL` to make it clear which API it is represents. 

The `ForgotPasswordCode` test has been fixed (I think). The mock handler was changed to use a regex to match against the API path.

The `__API_BASE_URL__` variable has been removed and replaced with `import.meta.env.VITE_HUU_API_BASE_URL`. @Joshua-Douglas  had a good idea to keep using the documented way of accessing the environment variables. The custom code to ensure that the variable exists and has been assigned a URL still remains in `vite.config.ts`

More documentation has been added to explain how the app is configured with environment variables and/or `.env` files.

Please have a look again. Thank you!

**Comment by erikguntner** [comment]
@paulespinosa I updated the variable name!

**Comment by paulespinosa** [comment]
> @paulespinosa I updated the variable name!

@erikguntner thank you! I re-ran the action and it succeeded :rocket:. Before we continue, a question about the E2E test. When the E2E test is run, the app is configured to connect to the HUU API at `https://dev.homeunite.us/api/` https://github.com/hackforla/HomeUniteUs/actions/runs/6065153640/job/16493059934#step:7:11. Question: during the E2E test, are real requests being made to that URL or are the requests intercepted by the test harness?

**Comment by paulespinosa** [comment]
The `build-deploy-ec2.yaml` action was manually run against this PR's branch. It builds and deploys successfully. We can see the action's results at https://github.com/hackforla/HomeUniteUs/actions/runs/6092098958

1. The version has been verified. The git commit on the swagger UI matches the latest commit of this branch: **0490241**
pic-selected-230906-0952-58

pic-selected-230906-0953-16

2. We see that the API url is configured in the app and it is possible to login.
pic-selected-230906-1003-50

**Review by Joshua-Douglas — APPROVED** [review]
Great changes @paulespinosa, LGTM! Sorry I didn't approve sooner.

**Comment by paulespinosa** [comment]
Hi @erikguntner and @Joshua-Douglas Sorry for another PR request! I encountered non-deterministic testing errors (random failures) and incorrect test implementation. So, the tests for `ForgotPasswordCode.test.tsx` and `ResetPassword.test.tsx` have been updated. 

Along with the changes were a refactor of the GitHub workflows: 
* One workflow just runs tests: Runs on PRs and pushes to main.
* One workflow runs tests, builds, and deploys: Runs on manual execution of the workflow.

I checked the deploy still works with the app containing the correct URL.

**Review by erikguntner — APPROVED** [review]
@paulespinosa awesome job getting those tests working! This all looks good to me.

File: app/vite.config.ts [hunk 2]
```diff
@@ -1,25 +1,62 @@
...
+      ),
+    },
+    plugins: [react()],
+    server: {
+      port: 4040,
+      proxy: {
+        '/api': {
+          target: 'http://localhost:8080',
+          changeOrigin: true,
+          secure: false,
```
Comment (Joshua-Douglas) [inline app/vite.config.ts:52]: This is a development-only server proxy, right? Just want to make sure that allowing unsecured proxies does not pose a security risk to the production application. 

I believe we use the VITE server for development, and `nginx` for production right?

**Review by Joshua-Douglas — APPROVED** [review]
Hey @paulespinosa, 

Changes look good:
* Great write up!
* The regex edit you made appears correct to me (it should match any string that ends in "api/auth/forgot_password")
* Removing async keyword from test looks correct. I don't see any async functions within the relevant test setup and execution.

I did have one question, but I don't see any issues. Approved, and we are good to merge unless you think anything is missing.

File: app/vite.config.ts (hunk 2 above)
Comment (paulespinosa) [reply app/vite.config.ts:52]: @Joshua-Douglas yes development only. The production front-end is compiled into html, JS, css files and served by Nginx.

Comment (paulespinosa) [reply app/vite.config.ts:52]: A static site.

**Pull request merged**